### ISSUE TITLE
torch.nn.DataParallel result differs from single GPU

### ISSUE THREAD

**Issue by Godricly** [pytorch/pytorch#24391]
## 🐛 Bug
the outputs for torch.nn.DataParallel is different from single GPU.

## To Reproduce
```
import numpy as np
import torch
from torch import nn
import torch.nn.init as init

def init_weights(modules):
    for m in modules:
        if isinstance(m, nn.Conv2d):
            init.xavier_uniform_(m.weight.data)
            if m.bias is not None:
                m.bias.data.zero_()
        elif isinstance(m, nn.BatchNorm2d):
            m.weight.data.fill_(1)
            m.bias.data.zero_()
        elif isinstance(m, nn.Linear):
            m.weight.data.normal_(0, 0.01)
            m.bias.data.zero_()


class double_conv(nn.Module):
    def __init__(self, in_ch, mid_ch, out_ch):
        super(double_conv, self).__init__()
        self.conv = nn.Sequential(
            nn.Conv2d(in_ch, mid_ch, kernel_size=1),
            nn.BatchNorm2d(mid_ch),
            nn.ReLU(inplace=True),
            nn.Conv2d(mid_ch, out_ch, kernel_size=3, padding=1),
            nn.BatchNorm2d(out_ch),
            nn.ReLU(inplace=True)
        )

    def forward(self, x):
        x = self.conv(x)
        return x

def model():
    return double_conv(3,4,4)

if __name__ == '__main__':
    model1 = model()
    model2 = model()
    model3 = model()
    init_weights(model1.conv)
    model2.load_state_dict(model1.state_dict())
    model3.load_state_dict(model1.state_dict())
    dummy = torch.tensor(np.random.uniform(size=(2,3,64,64)).astype(np.float32))

    model1 = model1.to(0)
    model2 = model2.to(1)
    model3 = model3.to(1)
    model3 = torch.nn.DataParallel(model3, device_ids=[1,2])
    # model3 = torch.nn.DataParallel(model3, device_ids=[0,1])

    out1 = model1(dummy.cuda(0))
    out2 = model2(dummy.cuda(1))
    out3 = model3(dummy.cuda(1))

    diff1 = torch.abs(out1.cpu() - out2.cpu()).max()
    diff2 = torch.abs(out1.cpu() - out3.cpu()).max()
    print(diff1, diff2)
```

1. python to run
2. the diff 1 is zero , while diff2 is not

## Environment
Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti
GPU 8: GeForce RTX 2080 Ti
GPU 9: GeForce RTX 2080 Ti

Nvidia driver version: 418.39
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.17.0
[pip3] torch==1.2.0
[pip3] torchvision==0.4.0
[conda] Could not collect

**Comment by zou3519** [comment]
I believe that this is expected behavior. BatchNorm performs differently because DataParallel splits batches to be sent across GPUs so the result is different. 

You are probably looking for SyncBatchNorm: https://pytorch.org/docs/master/nn.html#torch.nn.SyncBatchNorm

**Comment by soumith** [comment]
closing based on Richard's comment. feel free to reopen the issue if you disagree.
